### PR TITLE
Check array contiguity in `copy`

### LINKED ISSUE #2061: Broken copy across CUDA devices with non trivial-strides
Copying arrays across CUDA devices with non-trivial strides is broken. The resulting arrays is wrong. 

Note that the bug occurs only if `cp.array` is called with the default order of `'K'` meaning the array on the original device 0 is never converted into a contiguous array before being transferred.

#### Reproduction code

```python
import cupy as cp

def to_noncont(x):
    slices = (slice(None, None, 2),) * x.ndim
    x_noncont = cp.zeros(tuple([s * 2 for s in x.shape]), dtype=x.dtype)
    x_noncont[slices] = x
    x_noncont = x_noncont[slices]
    assert x_noncont.shape == x.shape
    return x_noncont

with cp.cuda.Device(0):
    x = cp.arange(6).reshape(2, 3)
    x_noncont = to_noncont(x)

with cp.cuda.Device(1):
    x_noncont_broken = cp.array(x_noncont)

# x_noncont_broken:
# [[0 1 2]
#  [0 0 0]]
```

### PR DESCRIPTION
**Pull request by okuta**
Fix #2061 and #1534.

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -422,8 +422,13 @@ cdef class ndarray:
         finally:
             runtime.setDevice(dev_id)
         newarray = ndarray(x.shape, dtype=x.dtype)
+        if not x._c_contiguous and not x._f_contiguous:
+            raise NotImplementedError(
+                'CuPy cannot copy non-contiguous array between devices.')
```
Comment (hvy) [inline cupy/core/core.pyx]: Just to make sure, do you mean it's a "TODO", or actually that it's never going to be supported (in which case we should not use `NotImplementedError`)?

Comment (okuta) [reply cupy/core/core.pyx]: It means TODO. In some cases, we can implement it.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
LGTM!

**Pull request merged**